usr/lib/cca_stdll/cca_specific.c

```diff
@@ -2858,8 +2882,22 @@ static CK_RV ccatok_pkey_skey2pkey(STDLL_TokData_t *tokdata,
     pkey_wrap_handler_data.pkey_buflen_p = &pkey_buflen;
     pkey_wrap_handler_data.keytype = key_type;
 
-    ret = cca_iterate_adapters(tokdata, ccatok_pkey_sec2prot,
-                               &pkey_wrap_handler_data);
+    while (num_retries < 3600) {
+
+        ret = cca_iterate_adapters(tokdata, ccatok_pkey_sec2prot,
+                                   &pkey_wrap_handler_data);
+
+        if (ret != CKR_OK || !pkey_wrap_handler_data.wrap_was_successful) {
+
+            if (ccatok_pkey_token_is_encr_with_new_mk(tokdata, key_type, mkvp)) {
+                if (ccatok_pkey_mk_change_active(tokdata, key_type)) {
+                    sleep(1);
+                    num_retries++;
+                    continue;
+                }
+            }
+        }
```

Comment:
You are missing to break the loop if pkey_wrap_handler_data.wrap_was_successful was true. You don't want to loop if the conversion was successful.

Reply:
I guess you can make this commit non-experimental now. The code looks good, so you can merge it into commit "CCA pkey option: determine firmware wkvp" or keep it as separate commit (but then with changed commit title).

Reply:
BTW: you always created macros to handle concurrent MK change ops. Should we also introduce two macros here, like BEGIN_RETRY_SEC2PROT / END_RETRY_SEC2PROT ?

Reply:
I created such macros for cases where the same logic is needed at many places. 
This here is needed at exact this one and only place, so no need to create a macro.

Actually RETRY_NEW_MK_BLOB_START / RETRY_NEW_MK_BLOB_END is a macro for that matter, but it does not fit here, because you are dealing with the pkey ioctl, not with CCA verbs.